### PR TITLE
feat: monitor debug curl

### LINKED ISSUE #5116: Insert/Copy curl of request
### 📑 I have found these related issues/pull requests

no

### 🏷️ Feature Request Type

New monitor, Change to existing monitor

### 🔖 Feature description

Add possibility to insert new monitor as curl and add possibility to copy curl of request.

### ✔️ Solution

Add possibility to insert new monitor as curl and add possibility to copy curl of request.

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by nekojanai**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Fixes #5116

comments appreciated

## Type of change

Please delete any options that are not relevant.

- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

|  |  |
|--------|--------|
| uptime-kuma-new-debug-button | outdated uptime-kuma-new-debug-dialog |
| | 
uptime-kuma-new-debug-dialog-update |

### CODE REVIEW

File: src/pages/EditMonitor.vue [hunk 1]
```diff
@@ -1000,9 +1010,20 @@
             <RemoteBrowserDialog ref="remoteBrowserDialog" />
         </div>
     </transition>
+    <div ref="modal" class="modal fade" tabindex="-1">
+        <div class="modal-dialog modal-dialog-centered">
+            <div class="modal-content">
+                <div class="modal-body">
+                    <textarea id="curlDebug" v-model="curlCommand" class="form-control mb-3" placeholder="curl -v 'https://example.com/healthcheck'" readonly></textarea>
+                    <button id="debug-copy-btn" class="btn btn-primary" type="button" @click.stop="copyToClipboard">{{ $t("Copy") }}</button>
```
Comment (CommanderStorm) [inline src/pages/EditMonitor.vue:1018]: I don't quite like this button placement (it somehow seems a bit off, but I can't quite put my finger on it..). Experimenting with different ways of doing this have not led me to an immediate better solution.

<details><summary>Maybe a copy button in the top right might be bettter. What do you think? (click to expand)</summary>
<p>

image

</p>
</details>

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Overall, this looks very good.
Sorry for stepping on your toes in https://github.com/louislam/uptime-kuma/issues/5116, that was not intended ^^

I have left a few comments below. While they might seem like a lot, I think they should be fairly straightforeward to implement ^^
Please have a look at https://github.com/louislam/uptime-kuma/pull/5152#discussion_r1785149705 first to make your live simpler while implmenting the changes.

File: src/pages/EditMonitor.vue [hunk 2]
```diff
@@ -1108,6 +1130,47 @@ export default {
 
     computed: {
 
+        curlCommand() {
+            let method = "";
+            let headers = "";
+            let body = "";
+            let basicAuth = "";
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1137]: This is such a good suggestions, it really does make things so much easier, thanks!

File: src/pages/EditMonitor.vue [hunk 3]
```diff
@@ -1108,6 +1130,47 @@ export default {
...
+                } catch (e) {
+                    json = this.monitor.body.replace(/\n/g, "");
+                    body = ` --json '${json}'`;
+                }
+                method = "";
+            } else if (this.monitor.body && this.monitor.method === "POST" && this.monitor.httpBodyEncoding === "xml") {
+                headers += " -H 'Content-Type: application/xml'";
+                body = ` -d ${this.monitor.body.replace(/\n/g, "")}`;
+            } else if (this.monitor.body) {
+                body = ` -d ${this.monitor.body.replace(/\n/g, "")}`;
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1166]: As that one comment on that post said, the triple quotes dont't really do anything...

[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/pages/EditMonitor.vue:1166]: Fuck


That really makes things more difficult.

File: src/pages/EditMonitor.vue [hunk 4]
```diff
@@ -1108,6 +1130,47 @@ export default {
...
+            }
+            if (this.monitor.authMethod === "basic") {
+                basicAuth = ` -u ${this.monitor.basic_auth_user}:${this.monitor.basic_auth_pass}`;
+            }
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1170]: `--oauth-bearer` does not exist in the ui, there's an option that performs an oauth client credential flow but that does the whole round trip of getting the bearer token and then attaches it to the request so there's no way to just specify the token, maybe a good first issue to create?

File: src/pages/EditMonitor.vue [hunk 5]
```diff
@@ -1108,6 +1130,47 @@ export default {
...
+                    for (const [ key, value ] of Object.entries(jsonParsed)) {
+                        headers += ` -H '${key}: ${value}'`;
+                    }
+                } catch (e) {
+                    headers = ` -H '${this.monitor.headers.replace(/\n/g, "")}'`;
+                }
+            }
+            if (this.monitor.body && this.monitor.httpBodyEncoding === "json") {
+                let json = "";
+                try {
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1153]: Exactly, I was trying to trim whitespace if the supplied data is correctly formatted json.

File: src/pages/EditMonitor.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/pages/EditMonitor.vue:1170]: Tokens in oauth are usually quite short-lived as far as I rememebr, right?
If that is so, specifying a token likely does not achive the goal of monitoring something.

=> lets just add a warning that tells the user that the oauth client credential flow is not supported in curl, right?

File: src/pages/EditMonitor.vue [hunk 6]
```diff
@@ -1108,6 +1130,47 @@ export default {
 
     computed: {
 
+        curlCommand() {
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1133]: There's no way to specify the user agent in the current ui.
I added `--insecure`.
I now also added `--location --max-redirs <number>`
And I added `--max-time` and `--retry <number>`

Proxy support is out of scope for this pr imo.

File: src/pages/EditMonitor.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue]: eh, I don't think so anymore, contrary to my initial assumption, if all multiline text blocks are at least quoted once the command should be handled well by most shells?

File: src/pages/EditMonitor.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1172]: Yep, sounds good!

File: src/pages/EditMonitor.vue (hunk 6 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/pages/EditMonitor.vue:1133]: > There's no way to specify the user agent in the current

The user agent is automatically set to the following by us to
https://github.com/louislam/uptime-kuma/blob/da168fc22082a4bdeae313ba90e0596366ecf87b/server/uptime-kuma-server.js#L488

> Proxy support is out of scope

Please add a warning in the modal if the user has a proxy configured.
Small warnings what is currently not supported go a long way 😅

File: src/pages/EditMonitor.vue (hunk 1 above)
Comment (nekojanai) [reply src/pages/EditMonitor.vue]: I like that idea, implemented.

File: src/pages/EditMonitor.vue [hunk 7]
```diff
@@ -1000,9 +1010,20 @@
...
     </transition>
+    <div ref="modal" class="modal fade" tabindex="-1">
+        <div class="modal-dialog modal-dialog-centered">
+            <div class="modal-content">
+                <div class="modal-body">
+                    <textarea id="curlDebug" v-model="curlCommand" class="form-control mb-3" placeholder="curl -v 'https://example.com/healthcheck'" readonly></textarea>
+                    <button id="debug-copy-btn" class="btn btn-primary" type="button" @click.stop="copyToClipboard">{{ $t("Copy") }}</button>
+                </div>
+            </div>
+        </div>
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue]: I placed some next just now, let's see if that's sufficient, unfortunately component interpolation does not let me add newlines?

[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/pages/EditMonitor.vue]: It does, via `<br/>` (requires its own template), but I would suggest using two translations. This way, translations are also better (?)

File: src/pages/EditMonitor.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (CommanderStorm) [reply src/pages/EditMonitor.vue]: 🤔 the following seems to work too. Should we go with that or with the simple `"..."` version?

image

File: src/pages/EditMonitor.vue [hunk 8]
```diff
@@ -1108,6 +1138,53 @@ export default {
...
+                let json = "";
+                try {
+                    // trying to parse the supplied data as json to trim whitespace
+                    json = JSON.stringify(JSON.parse(this.monitor.body));
+                } catch (e) {
+                    json = this.monitor.body;
+                }
+                command.push("--header", "'Content-Type: application/json'", "\\\n", "--data", `'${json}'`, "\\\n");
+            } else if (this.monitor.body && this.monitor.httpBodyEncoding === "xml") {
+                command.push("--headers", "'Content-Type: application/xml'", "\\\n", "--data", `'${this.monitor.body}''`, "\\\n");
```
Comment (CommanderStorm) [inline src/pages/EditMonitor.vue:1173]:
```suggestion
                command.push("--headers", "'Content-Type: application/xml'", "\\\n", "--data", `'${this.monitor.body}'`, "\\\n");
```

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
The command generation part looks indeed much better.
=> No complaints, thanks

I have left two styling requests below regarding the helptexts and alerts. They currently don't quite look finished, but that should be a simple fix

File: src/pages/EditMonitor.vue [hunk 9]
```diff
@@ -1018,7 +1018,10 @@
                     <button id="debug-copy-btn" class="btn btn-outline-primary position-absolute top-0 end-0 mt-3 me-3 border-0" type="button" @click.stop="copyToClipboard">
                         <font-awesome-icon icon="copy" />
                     </button>
-                    <p>{{ $t('CurlDebugInfo') }}</p>
+                    <p>{{ $t('CurlDebugInfoLine0') }}</p>
+                    <p>{{ $t('CurlDebugInfoLine1') }}</p>
```
[the remark this answers is not in the document]
Comment (nekojanai) [reply src/pages/EditMonitor.vue:1022]: There was no specific reason, I corrected it now.

File: src/lang/en.json [hunk 10]
```diff
@@ -249,6 +251,14 @@
     "PushUrl": "Push URL",
     "HeadersInvalidFormat": "The request headers are not valid JSON: ",
     "BodyInvalidFormat": "The request body is not valid JSON: ",
+    "CopyToClipboardError": "Couldn't copy to clipbard: {error}",
+    "CopyToClipboardSuccess": "Copied!",
+    "CurlDebugInfo": "To debug the monitor, you can either paste this into your own machines terminal or into the machines terminal which uptime kuma is running on and see what you are requesting.{0}Please be aware of networking differences like {1}, {2} or {3}.",
+    "CurlDebugInfoFirewalls": "firewalls",
+    "CurlDebugInfoDnsResolvers": "dns resolvers",
+    "CurlDebugInfoDockerNetworks": "docker networks",
```
Comment (CommanderStorm) [inline src/lang/en.json:259]:
```suggestion
    "firewalls": "firewalls",
    "dns resolvers": "dns resolvers",
    "docker networks": "docker networks",
```

File: src/pages/EditMonitor.vue [hunk 11]
```diff
@@ -1000,9 +1010,42 @@
...
+                <div class="modal-body">
+                    <textarea id="curl-debug" v-model="curlCommand" class="form-control mb-3" readonly wrap="off"></textarea>
+                    <button id="debug-copy-btn" class="btn btn-outline-primary position-absolute top-0 end-0 mt-3 me-3 border-0" type="button" @click.stop="copyToClipboard">
+                        <font-awesome-icon icon="copy" />
+                    </button>
+                    <i18n-t keypath="CurlDebugInfo" tag="p" class="form-text">
+                        <br>
+                        <a href="https://xkcd.com/2259/" target="_blank">{{ $t('CurlDebugInfoFirewalls') }}</a>
+                        <a href="https://www.reddit.com/r/sysadmin/comments/rxho93/thank_you_for_the_running_its_always_dns_joke_its/" target="_blank">{{ $t('CurlDebugInfoDnsResolvers') }}</a>
+                        <a href="https://youtu.be/bKFMS5C4CG0" target="_blank">{{ $t('CurlDebugInfoDockerNetworks') }}</a>
```
Comment (CommanderStorm) [inline src/pages/EditMonitor.vue:1025]:
```suggestion
                        <template #newiline>
                            <br>
                        </template>
                        <template #firewalls>
                            <a href="https://xkcd.com/2259/" target="_blank">{{ $t('firewalls') }}</a>
                        </template>
                        <template #dns_resolvers>
                            <a href="https://www.reddit.com/r/sysadmin/comments/rxho93/thank_you_for_the_running_its_always_dns_joke_its/" target="_blank">{{ $t('dns resolvers') }}</a>
                        </template>
                        <template #docker_networks>
                            <a href="https://youtu.be/bKFMS5C4CG0" target="_blank">{{ $t('docker networks') }}</a>
                        </template>
```

File: src/lang/en.json [hunk 12]
```diff
@@ -249,6 +251,14 @@
     "PushUrl": "Push URL",
     "HeadersInvalidFormat": "The request headers are not valid JSON: ",
     "BodyInvalidFormat": "The request body is not valid JSON: ",
+    "CopyToClipboardError": "Couldn't copy to clipbard: {error}",
+    "CopyToClipboardSuccess": "Copied!",
+    "CurlDebugInfo": "To debug the monitor, you can either paste this into your own machines terminal or into the machines terminal which uptime kuma is running on and see what you are requesting.{0}Please be aware of networking differences like {1}, {2} or {3}.",
```
Comment (CommanderStorm) [inline src/lang/en.json:256]:
```suggestion
    "CurlDebugInfo": "To debug the monitor, you can either paste this into your own machines terminal or into the machines terminal which uptime kuma is running on and see what you are requesting.{newiline}Please be aware of networking differences like {firewalls}, {dns_resolvers} or {docker_networks}.",
```

File: src/pages/EditMonitor.vue [hunk 13]
```diff
@@ -1000,9 +1010,42 @@
...
+                    </i18n-t>
+                    <div v-if="monitor.authMethod === 'oauth2-cc'" class="alert alert-warning d-flex align-items-center gap-2" role="alert">
+                        <div role="img" aria-label="Warning:">⚠️</div>
+                        <i18n-t keypath="CurlDebugInfoOAuth2CCUnsupported" tag="div">
+                            <code>curl</code>
+                            <br>
+                            <code>--oauth2-bearer TOKEN</code>
```
Comment (CommanderStorm) [inline src/pages/EditMonitor.vue:1032]:
```suggestion
                            <template #curl>
                                <code>curl</code>
                            </template>
                            <template #newline>
                                <br>
                            </template>
                            <template #oauth2_bearer>
                                <code>--oauth2-bearer TOKEN</code>
                            </template>
```

File: src/pages/EditMonitor.vue [hunk 14]
```diff
@@ -1000,9 +1010,42 @@
...
+                        </i18n-t>
+                    </div>
+                    <div v-if="monitor.proxyId" class="alert alert-warning d-flex align-items-center gap-2" role="alert">
+                        <div role="img" aria-label="Warning:">⚠️</div>
+                        <i18n-t keypath="CurlDebugInfoProxiesUnsupported" tag="div">
+                            <code>curl</code>
```
Comment (CommanderStorm) [inline src/pages/EditMonitor.vue:1038]:
```suggestion
                            <template #curl>
                                <code>curl</code>
                            </template>
```

File: src/lang/en.json [hunk 15]
```diff
@@ -249,6 +251,14 @@
...
+    "CurlDebugInfoFirewalls": "firewalls",
+    "CurlDebugInfoDnsResolvers": "dns resolvers",
+    "CurlDebugInfoDockerNetworks": "docker networks",
+    "CurlDebugInfoOAuth2CCUnsupported": "Full oauth client credential flow is not supported in {0}.{1}Please get a bearer token and pass it via the {2} option.",
+    "CurlDebugInfoProxiesUnsupported": "Proxy support in the above {0} command is currently not implemented.",
```
Comment (CommanderStorm) [inline src/lang/en.json:261]:
```suggestion
    "CurlDebugInfoOAuth2CCUnsupported": "Full oauth client credential flow is not supported in {curl}.{newline}Please get a bearer token and pass it via the {oauth2_bearer} option.",
    "CurlDebugInfoProxiesUnsupported": "Proxy support in the above {curl} command is currently not implemented.",
```

**Review by CommanderStorm — APPROVED** [review]
Thanks for debugging functionality! 🎉 
(The changes below are stylistic to make translating hopefully a bit simpler)



> [!NOTE]
> This PR is part of the [`v2.0` merge window](https://github.com/louislam/uptime-kuma/milestone/24) => see #4500 for the bugs that need to be addressed before we can ship this release ^^

<sub>All changes in this PR are small and uncontroversial ⇒ merging with junior maintainer approval</sub>

**Comment by louislam** [comment]
image

I think all values should be escaped for security/safe.

**Comment by louislam** [comment]
Two packages are suggested by ChatGPT:

- execa.command (from Execa)
- shell-escape (simple, but a bit old)

**Pull request merged**